### PR TITLE
[R] Redesigned `xgboost()` interface skeleton

### PR DESCRIPTION
**Pull request by david-cortes**
ref https://github.com/dmlc/xgboost/issues/9810

This PR starts off the redesigned interface for `xgboost()`, by adding a bare-bones version that replaces the old one:
* Uses x/y interface instead of `xgb.DMatrix`.
* Enforces a match between the type and shape of `y` and the objective.
* Makes the `objective` optional, picking a default according to the type of `y`.
* Allows column name matching in parameters that refer to columns.
* Uses `xgb.QuantileDMatrix` whenever possible.
* Keeps extra metadata about `y` (column names and factor levels).
* Inherits from `xgb.Booster`, but has an additional class `xgboost`, so that the current methods can be used on it.

But:
* Doesn't yet make use of the additional metadata that it keeps in other methods, beyond the new `print` that's added here.
* Doesn't yet have a more idiomatic `predict` method.
* Has not yet updated examples and guides to use this function.
* Doesn't yet deal with the unidiomatic passing of `params`.
* Doesn't yet offer some functionalities that it should, like custom metrics.
* Has not yet updated the serialization doc for this new class.

Among others. It should nevertheless provide a base to work on.

A couple items I am not so sure about - would be helpful to get reviews on these:
* Does the function in this PR correctly capture all the possible cases in which there can be more than one score per leaf? (e.g. multinomial, multi-task).
* Is it treating `Surv` objects correctly for all cases?
* Where can I find reference papers to add to the docs explaining how XGBoost implements things like monotonicity constraints and interaction contrstraints? I see there's the original paper that introduces XGBoost that is mentioned in the docs, but I imagined that a lot more might have been written by the same authors over time.

CCing @mayer79 for review.

### CODE REVIEW

**Comment by trivialfis** [comment]
Thank you for sharing the early work. I will look into the new example and introduction for learning about the new interface.

**Comment by david-cortes** [comment]
> Thank you for sharing the early work. I will look into the new example and introduction for learning about the new interface.

To clarify: there's no intro and only a very small example, since there's no specialized `predict` yet for this class, for example.

**Comment by trivialfis** [comment]
Would you like me to add a feature branch in xgboost for you to iterate on?

**Comment by david-cortes** [comment]
> Would you like me to add a feature branch in xgboost for you to iterate on?

Thanks but I don't think there's any need for that. Or are you planning a 2.1.1 release soon?

File: R-package/R/xgboost.R
Comment (mayer79) [inline R-package/R/xgboost.R:304]: Instead of the ifelse() maybe

lb <- ub <- y[, 1L]
lb[y[, 2L] == 0] <- 0

File: R-package/R/xgboost.R
Comment (mayer79) [inline R-package/R/xgboost.R:315]: Maybe

```r
lb <- ub <- y[, 1L]
ub[y[, 2L] == 0] <- Inf
```

File: R-package/R/xgboost.R
Comment (mayer79) [inline R-package/R/xgboost.R:504]:
```suggestion
    nthreads <- nthreads[1L]
```

File: R-package/R/xgboost.R
Comment (mayer79) [inline R-package/R/xgboost.R:506]:
```suggestion
  if (is.na(nthreads) || nthreads < 1L) {
```

File: R-package/R/xgboost.R
Comment (mayer79) [inline R-package/R/xgboost.R:574]:
```suggestion
      if (length(matched) > 0L && length(matched) < length(feature_weights)) {
```

File: R-package/R/xgboost.R
Comment (mayer79) [inline R-package/R/xgboost.R:601]: We need the names multiple times
```r
nms <- names(monotone_constraints)
```

File: R-package/R/xgboost.R
Comment (mayer79) [inline R-package/R/xgboost.R:622]:
```suggestion
          paste(utils::head(names(monotone_constraints)), collapse = ", ")
```

File: R-package/R/xgboost.R
Comment (mayer79) [inline R-package/R/xgboost.R:697]:
```suggestion
            paste(utils::head(idx[which(is.na(out))]), collapse = ", ")
```

File: R-package/R/xgboost.R
Comment (mayer79) [inline R-package/R/xgboost.R:715]:
```suggestion
              paste(utils::head(idx[which(idx != floor(idx))]), collapse = ", ")
```

File: R-package/R/xgboost.R
Comment (mayer79) [inline R-package/R/xgboost.R:577]:
```suggestion
          utils::head(setdiff(colnames(x), names(feature_weights)))
```

File: R-package/R/xgboost.R
Comment (mayer79) [inline R-package/R/xgboost.R:614]:
```suggestion
            names(monotone_constraints)[duplicated(names(monotone_constraints))] |> utils::head(),
```

File: R-package/R/xgboost.R
Comment (mayer79) [inline R-package/R/xgboost.R:974]:
```suggestion
    v_sub <- utils::head(v, 5L)
```

**Review by mayer79 — APPROVED** [review]
Awesome work @david-cortes !

The auto-selection of the objective is really neat.

While reading the code, I added some detail stuff (not that they are important, it was just to keep me busy).

Some additional comments:

- I'd prefer helper functions like `process.x.and.col.args()` to be named like `_process_x_and_col_args()`.
- We will also need to deal with `X` (store names of features, store factor levels of character/factor columns).

File: R-package/R/xgboost.R
Comment (david-cortes) [reply R-package/R/xgboost.R:490]: This is already checked in the core library.

File: R-package/R/xgboost.R
Comment (david-cortes) [reply R-package/R/xgboost.R:356]: It is not necessarily the case that `NROW` will return an integer type, so there isn't always an advantage in declaring the comparison as integer.

File: R-package/R/xgboost.R
Comment (david-cortes) [reply R-package/R/xgboost.R:601]: Retrieving the names does not generate a deep copy of them.

File: R-package/R/xgboost.R
Comment (mayer79) [reply R-package/R/xgboost.R:356]: > It is not necessarily the case that `NROW` will return an integer type, so there isn't always an advantage in declaring the comparison as integer.

It either returns length() or dim()[1], both are ints. But we are discussing nano seconds ;).

Comment (david-cortes) [reply R-package/R/xgboost.R:356]:
```r
class(NROW(1:3000000000))
```
```
[1] "numeric"
```

Comment (mayer79) [reply R-package/R/xgboost.R:356]: I see! But that's quite a large number of quantiles to model ;-).

File: R-package/R/xgboost.R
Comment (trivialfis) [inline R-package/R/xgboost.R:257]: Nice change for survival training!

File: R-package/R/xgboost.R
Comment (trivialfis) [inline R-package/R/xgboost.R:327]: Could you please add some comments for these indexing?

File: R-package/R/xgboost.R
Comment (trivialfis) [inline R-package/R/xgboost.R:497]: Is this going to make single thread the default for R xgboost?

File: R-package/R/xgboost.R
Comment (trivialfis) [inline R-package/R/xgboost.R:506]: XGBoost takes 0 as using all available cores.

File: R-package/R/xgboost.R
Comment (trivialfis) [inline R-package/R/xgboost.R:518]: We can support COO internally, for CSC, maybe we can simply transpose in R. Do you think it can help the make code less complex?

File: R-package/R/xgboost.R
Comment (trivialfis) [inline R-package/R/xgboost.R:551]: Do we need to worry about `data.table` and potentially other libraries that inherit from these known types in this error message?

File: R-package/R/xgboost.R
Comment (trivialfis) [inline R-package/R/xgboost.R:567]: Is it possible that we can use the check in c++ https://github.com/dmlc/xgboost/blob/513d7a7d84678fc4e75c18899fe52263258f0290/src/data/data.cc#L785 ?

File: R-package/R/xgboost.R
Comment (david-cortes) [reply R-package/R/xgboost.R:22]: They are meant to be passed as arguments to the function `xgboost()` instead of being passed through `params`.

File: R-package/R/xgboost.R
Comment (david-cortes) [reply R-package/R/xgboost.R]: Changed the condition and added it to the docs.

File: R-package/R/xgboost.R
Comment (david-cortes) [reply R-package/R/xgboost.R:495]: No, the default is to use all threads, as obtained when calling `parallel::detectCores()`.

File: R-package/R/xgboost.R
Comment (david-cortes) [reply R-package/R/xgboost.R:325]: Added a comment block above.

File: R-package/R/xgboost.R
Comment (david-cortes) [reply R-package/R/xgboost.R:549]: No, because all `data.table` objects are also `data.frame`s.
```r
class(data.table())
````
```
[1] "data.table" "data.frame"
```

File: R-package/R/xgboost.R
Comment (david-cortes) [reply R-package/R/xgboost.R:516]: Maybe, but what would be the benefit in such case if the data still needs to be duplicated?

Comment (david-cortes) [reply R-package/R/xgboost.R]: Modified it to accept COO and CSC for QDM, with a casting for now until direct QDM construction from COO is added.

File: R-package/R/xgboost.R
Comment (trivialfis) [reply R-package/R/xgboost.R:22]: `monotone_constraints` and `interaction_constraints` might be better as a part of normal params? This way it feels more consistent. It's difficult from a user's perspective to understand why some of these parameters are different.

File: R-package/R/xgboost.R
Comment (trivialfis) [reply R-package/R/xgboost.R:547]: I'm aware of this, hence the question, is it possible that this error message might lead the user to think `data.frame` is not supported? Or it's common knowledge in the R world?

Comment (david-cortes) [reply R-package/R/xgboost.R:547]: No, because if `class(x)` contains `data.frame` as an entry, `inherits` will return `TRUE`, so there will be no error message in a derived class like `data.table` or `tibble`. And if it is a class like `arrow` or polars `DataFrame` which do not inherit from `data.frame`, then the error will show as expected, since they are not supported.

File: R-package/R/xgboost.R
Comment (david-cortes) [reply R-package/R/xgboost.R:22]: From my perspective:
* Some parameters I chose to put as function arguments in order to give them a default value visible to the user (e.g. `nthreads`) or make the default value change according to the data (e.g. `objective`).
* Some parameters are independent of the data (e.g. regularization) or work regardless of whether the data supports them or not (e.g. max categ. to one-hot), while others refer specifically to columns in the data (e.g. interaction constraints) and would not work with a different dataset.
* Ideally, later on there should be a function like `xgb.control` or similar that would list the parameters and offer auto-completion by having them as function arguments (e.g. like `VGAM::vglm.control` or `glmx::glmx.control`). Those train-control parameters typically involve things that are independent of the data.
* Alternatively, and as is more common in R, _all_ parameters could be moved to function arguments (e.g. like `ranger::ranger` or `glmnet::glmnet`), but perhaps that would make the docs very hard to use.

> It's difficult from a user's perspective to understand why some of these parameters are different.

Yes, it's a bit arbitrary, but that is also the situation in the other interfaces - e.g. the number of boosting rounds is a function argument, as is `verbose_eval` in the python interface. It also feels like parameters such as `seed` should be in the train control list, but that would have other side effects due to the logic introduced in an earlier PR (https://github.com/dmlc/xgboost/pull/10029).

Comment (david-cortes) [reply R-package/R/xgboost.R:22]: Perhaps for v3 you could consider having a division between data-dependent parameters (e.g. `feature_weights` is passed under `DMatrix`, while `monotone_contraints` is passed under `params`) and data-independent parameters.

Comment (trivialfis) [reply R-package/R/xgboost.R:22]: Thank you for the detailed explanation. Please think of me as a newcomer instead of a maintainer. Or think of the future GitHub issues we will need to reply to when someone puts the parameter in the wrong place and asks why.

>  but that is also the situation in the other interfaces

I think that's only true for the low-level interface, and some of them are only in this way because we think of it as an unnecessary breaking change. For higher-level interfaces like the scikit-learn, or Spark, they have a relatively consistent interface. For the scikit-learn, we are moving the parameters according to the scikit-learn estimator guideline where the constructor takes hyper-parameter and sometimes feature-related data like feature types, while the `fit` method takes data(sample)-dependent parameters that can be split for early stopping, etc.

> default value visible to the user

We can compromise on this, maybe the document is sufficient?

> Alternatively, and as is more common in R, all parameters could be moved to function arguments

That sounds reasonable to me. Personally, I prefer consistency, which has been a pain point for XGBoost for a while now. Would be great to see improvement in this area.

> Perhaps for v3 you could consider having a division between data-dependent parameters (e

Hmm, since this is a new interface already, maybe it's better to do it right the first time if we know where the issue is?

Comment (david-cortes) [reply R-package/R/xgboost.R:22]: Ok then, I'll move all parameters to function arguments, leaving them as undocumented `...` for now (to be changed in a later PR). We'll see how the docs and examples end up looking later on.

Comment (david-cortes) [reply R-package/R/xgboost.R]: Modified it by removing `params` and making all parameters as function arguments, by using `...` for now.

**Review by trivialfis — APPROVED** [review]

**Comment by trivialfis** [comment]
Thank you for the excellent work, will merge once the CI passes.

**Pull request merged**
